File: src/lib/sender.js
// Runs a batch of personalized SMS sends and reports progress via a simple
// subscribe callback (no server/polling needed - everything runs in-tab).

import { AdbError, sendSms } from "./adb-client.js";
import { normalizePhone } from "./excel.js";
import { render } from "./templating.js";

const DEFAULT_DELAY_SECONDS = 4;

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class SendJob {
  constructor(rows, template, phoneColumn, config) {
    this.rows = rows;
    this.template = template;
    this.phoneColumn = phoneColumn;
    this.config = config;
    this.status = "pending"; // pending -> running -> done | cancelled
    this.results = [];
    this.cancelRequested = false;
    this._listeners = new Set();
  }

  onUpdate(listener) {
    this._listeners.add(listener);
    return () => this._listeners.delete(listener);
  }

  _notify() {
    const snapshot = this.snapshot();
    for (const listener of this._listeners) listener(snapshot);
  }

  snapshot() {
    return {
      status: this.status,
      total: this.rows.length,
      completed: this.results.length,
      results: [...this.results],
    };
  }

  cancel() {
    this.cancelRequested = true;
  }

  async run() {
    this.status = "running";
    this._notify();

    const { adb, dryRun = false, delaySeconds = DEFAULT_DELAY_SECONDS, manualTap = null } = this.config;

    for (const row of this.rows) {
      if (this.cancelRequested) {
        this.status = "cancelled";
        this._notify();
        return;
      }

      const message = render(this.template, row);
      const number = normalizePhone(row[this.phoneColumn]);
      const entry = { number, message };

      if (!number) {
        entry.ok = false;
        entry.error = "מספר טלפון חסר או לא תקין";
      } else if (dryRun) {
        entry.ok = true;
        entry.dryRun = true;
      } else {
        try {
          await sendSms(adb, number, message, manualTap);
          entry.ok = true;
        } catch (err) {
          entry.ok = false;
          entry.error = err instanceof AdbError ? err.message : String(err?.message || err);
        }
      }

      this.results.push(entry);
      this._notify();

      if (!dryRun && !this.cancelRequested) {
        await sleep(delaySeconds * 1000);
      }
    }

    if (this.status === "running") {
      this.status = "done";
      this._notify();
    }
  }
}

File: sms_sender/sender.py
"""Runs a batch of personalized SMS sends in a background thread and tracks
progress so the web UI can poll for live status.
"""

import threading
import time
import uuid

from . import adb as adb_mod
from .excel_utils import normalize_phone
from .templating import render

DEFAULT_DELAY_SECONDS = 4


class Job:
    def __init__(self, job_id, rows, template, phone_column, config):
        self.id = job_id
        self.rows = rows
        self.template = template
        self.phone_column = phone_column
        self.config = config
        self.status = "pending"  # pending -> running -> done | cancelled
        self.results = []
        self.cancel_requested = False
        self._lock = threading.Lock()

    def to_dict(self):
        with self._lock:
            results = list(self.results)
        return {
            "id": self.id,
            "status": self.status,
            "total": len(self.rows),
            "completed": len(results),
            "results": results,
        }

    def _append_result(self, entry):
        with self._lock:
            self.results.append(entry)


_JOBS = {}
_JOBS_LOCK = threading.Lock()


def create_job(rows, template, phone_column, config):
    job_id = uuid.uuid4().hex[:12]
    job = Job(job_id, rows, template, phone_column, config)
    with _JOBS_LOCK:
        _JOBS[job_id] = job
    return job


def get_job(job_id):
    with _JOBS_LOCK:
        return _JOBS.get(job_id)


def cancel_job(job_id):
    job = get_job(job_id)
    if job:
        job.cancel_requested = True


def _run_job(job):
    job.status = "running"
    serial = job.config.get("serial")
    adb_path = job.config.get("adb_path", "adb")
    dry_run = bool(job.config.get("dry_run", False))
    delay_seconds = float(job.config.get("delay_seconds", DEFAULT_DELAY_SECONDS))
    manual_tap = job.config.get("manual_tap")

    for row in job.rows:
        if job.cancel_requested:
            job.status = "cancelled"
            return

        message = render(job.template, row)
        number = normalize_phone(row.get(job.phone_column, ""))
        entry = {"number": number, "message": message}

        if not number:
            entry["ok"] = False
            entry["error"] = "מספר טלפון חסר או לא תקין"
        elif dry_run:
            entry["ok"] = True
            entry["dry_run"] = True
        else:
            try:
                adb_mod.send_sms(serial, number, message, adb_path=adb_path, manual_tap=manual_tap)
                entry["ok"] = True
            except adb_mod.AdbError as exc:
                entry["ok"] = False
                entry["error"] = str(exc)

        job._append_result(entry)

        if not dry_run and not job.cancel_requested:
            time.sleep(delay_seconds)

    if job.status == "running":
        job.status = "done"


def start_job(job):
    thread = threading.Thread(target=_run_job, args=(job,), daemon=True)
    thread.start()
    return thread
